// @egret/pro-cli-proxy/src/index.ts
import * as engineManager from '@egret/engine-manager';
import * as fs from 'fs';
import * as path from 'path';
import stripJsonComments from "strip-json-comments";

let projectRoot: string;

export function init(root: string) {
    projectRoot = root;
}

export function getEgretProCli() {
    const m = requireModule('./toolkit/pro-cli') as typeof import('../dist/pro-cli');
    return m;
}

export function createProject(destinationPath: string, options: import('../dist/pro-cli').CreateOptions) {
    const engineName = options.engineName;
    const engineRoot = getEngineRoot(engineName) as string;
    const m = requireModule('./toolkit/pro-cli', engineRoot) as typeof import('../dist/pro-cli');
    m.create(destinationPath, options);
}

export function getChildProcessCompiler() {
    const m = requireModule('./toolkit/child-process-compiler') as typeof import('../dist/child-process-compiler');
    return m;
}

export function getCompiler() {
    const m = requireModule('./toolkit/compiler') as typeof import('../dist/compiler');
    return m;
}

export function setEngineRoot(name: string, root: string) {
    engineManager.setEngineRoot(name, root);
}

export function getEngineRoot(name: string) {
    return engineManager.getEngineRoot(name);
}

export function getEngineList(): { name: string, path: string }[] {
    return engineManager.getEngineList();
}

function getCurrentEngineRoot() {
    const engineName = getEngineName();
    const engineRoot = getEngineRoot(engineName);
    if (!engineRoot) {
        throw new Error("未找到 engineRoot,请在 EgretPro 中进行设置");
    }
    return engineRoot;
}

function requireModule(name: string, egretProRoot?: string) {
    egretProRoot = egretProRoot || getCurrentEngineRoot();
    const cliPath = path.join(egretProRoot, name);
    const globalThis = global as any;
    const r = globalThis.require || eval('require');
    const m = r(cliPath);
    if (m.setEgretProRoot) {
        m.setEgretProRoot(egretProRoot);
    }
    return m;
}

function getEngineName() {
    const projectFilepath = path.join(projectRoot, 'egretpro.json');
    const content = fs.readFileSync(projectFilepath, 'utf-8');
    const json = JSON.parse(stripJsonComments(content));
    const engineName = json.engineName;
    if (!engineName) {
        throw new Error("未找到 engineName 字段,请在 egretpro.json 中设置 engineName 字段");
    }
    else {
        return engineName;
    }
}
